File: server/typescript/packages/render/test/extract/extract-conformance.test.ts
import { describe, test, expect } from "bun:test";
import { readFileSync, readdirSync, existsSync, statSync } from "node:fs";
import { join, dirname } from "node:path";
import { extract } from "../../src/extract/extract.js";
import {
  Format,
  FieldKind,
  scalar,
  enumField,
  enumArray,
  range,
  object,
  textContentField,
  type FieldSpec,
  type ExtractSchema,
} from "../../src/extract/types.js";
import type { NormalizeMode } from "../../src/extract/normalize.js";

// FR-010 cross-language extract-conformance corpus runner — the correctness gate.
// Each fixture dir under fixtures/extract-conformance/ holds:
//   schema.json   { "format": "JSON"|"XML", "rootName": "...", "fields": [...] }
//   input.txt     the raw (possibly dirty) LLM output
//   expected.json { "empty": bool, "states": { path: FieldExtraction }, "data": { field: value } }
// All cases must pass. The corpus is the oracle — do not weaken assertions.
// Mirrors ExtractConformanceTest.java / ExtractConformanceTests.cs exactly.

/** Walk up from this test dir to the repo root that contains fixtures/extract-conformance. */
function corpusRoot(): string {
  let dir = import.meta.dir;
  for (;;) {
    const candidate = join(dir, "fixtures", "extract-conformance");
    if (existsSync(candidate)) return candidate;
    const parent = dirname(dir);
    if (parent === dir) throw new Error("fixtures/extract-conformance not found");
    dir = parent;
  }
}

interface ExpectedJson {
  empty: boolean;
  states: Record<string, string>;
  data: Record<string, unknown>;
}

function parseFormat(s: string): Format {
  switch (s) {
    case "JSON":
      return Format.JSON;
    case "XML":
      return Format.XML;
    default:
      throw new Error(`Unknown format: ${s}`);
  }
}

function parseFieldKind(s: string): FieldKind {
  switch (s) {
    case "STRING":
      return FieldKind.STRING;
    case "INT":
      return FieldKind.INT;
    case "LONG":
      return FieldKind.LONG;
    case "DOUBLE":
      return FieldKind.DOUBLE;
    case "BOOLEAN":
      return FieldKind.BOOLEAN;
    case "ENUM":
      return FieldKind.ENUM;
    case "OBJECT":
      return FieldKind.OBJECT;
    default:
      throw new Error(`Unknown field kind: ${s}`);
  }
}

interface RawFieldJson {
  name: string;
  kind: string;
  required?: boolean;
  array?: boolean;
  enumValues?: string[];
  enumAlias?: Record<string, string>;
  // FR-011: enum coercion-pipeline schema keys.
  coerceDefault?: string;
  normalize?: string;
  default?: string;
  min?: number;
  max?: number;
  // @xmlText: scalar field that receives its element's text content (the #text sentinel).
  textContent?: boolean;
  // FR-011: nested-object sub-fields (present only for kind === "OBJECT").
  fields?: RawFieldJson[];
}

function parseNormalize(s: string | undefined): NormalizeMode {
  switch (s) {
    case undefined:
      return "strip"; // FR-011 global default
    case "none":
    case "collapse":
    case "strip":
      return s;
    default:
      throw new Error(`Unknown normalize mode: ${s}`);
  }
}

function parseField(f: RawFieldJson): FieldSpec {
  const name = f.name;
  const kind = parseFieldKind(f.kind);
  const required = f.required === true;

  if (kind === FieldKind.ENUM) {
    const build = f.array === true ? enumArray : enumField;
    return build(
      name,
      required,
      f.enumValues ?? [],
      f.enumAlias ?? {},
      f.coerceDefault ?? null,
      parseNormalize(f.normalize),
      f.default ?? null,
    );
  }
  if (kind === FieldKind.OBJECT) {
    const nested =
      f.fields === undefined
        ? null
        : { format: Format.JSON, rootName: name, fields: f.fields.map(parseField) };
    return object(name, required, f.array === true, nested);
  }
  if (f.min !== undefined || f.max !== undefined) {
    return range(name, kind, required, f.min ?? null, f.max ?? null);
  }
  // @xmlText: a scalar field that receives its element's text content (the #text sentinel).
  if (f.textContent === true) return textContentField(name, kind, required);
  // Phase B (generalized @default): a scalar `default` key fills an absent field, coerced to kind.
  return scalar(name, kind, required, f.default ?? null);
}

interface RawSchemaJson {
  format: string;
  rootName: string;
  // Optional per-fixture parse option: rootless XML (the response has no wrapper root element).
  rootless?: boolean;
  fields: RawFieldJson[];
}

function parseSchema(n: RawSchemaJson): ExtractSchema {
  return {
    format: parseFormat(n.format),
    rootName: n.rootName,
    fields: (n.fields ?? []).map(parseField),
  };
}

function assertCanonical(caseName: string, path: string, expected: unknown, actual: unknown): void {
  if (typeof expected === "number") {
    expect(typeof actual === "number" ? Math.abs(expected - (actual as number)) <= 1e-9 : false).toBe(true);
  } else {
    const expectedStr = typeof expected === "string" ? expected : String(expected);
    const actualStr = actual == null ? String(actual) : typeof actual === "string" ? actual : String(actual);
    expect(actualStr).toBe(expectedStr);
  }
}

/**
 * Flatten an assembled-data value into dotted leaf paths: objects recurse by key
 * (`prefix.key`), arrays recurse by index (`prefix[i]`), and every terminal scalar is recorded.
 * Mirrors the engine's per-field state enumeration so data leaves line up with state leaves.
 */
function flattenLeaves(prefix: string, value: unknown, out: Record<string, unknown>): void {
  if (Array.isArray(value)) {
    value.forEach((item, i) => flattenLeaves(`${prefix}[${i}]`, item, out));
  } else if (value !== null && typeof value === "object") {
    for (const [k, v] of Object.entries(value as Record<string, unknown>)) {
      flattenLeaves(prefix === "" ? k : `${prefix}.${k}`, v, out);
    }
  } else {
    out[prefix] = value;
  }
}

describe("extract-conformance corpus", () => {
  const corpus = corpusRoot();
  const cases = readdirSync(corpus)
    .filter((n) => statSync(join(corpus, n)).isDirectory())
    .filter((n) => existsSync(join(corpus, n, "schema.json")))
    .sort();

  expect(cases.length).toBe(31);

  for (const caseName of cases) {
    test(caseName, () => {
      const caseDir = join(corpus, caseName);
      const schemaNode = JSON.parse(readFileSync(join(caseDir, "schema.json"), "utf8")) as RawSchemaJson;
      const schema = parseSchema(schemaNode);
      const input = readFileSync(join(caseDir, "input.txt"), "utf8");
      const expected = JSON.parse(readFileSync(join(caseDir, "expected.json"), "utf8")) as ExpectedJson;

      // Optional per-fixture parse option: "rootless": true → the XML response has no wrapper root
      // element (the payload's fields ARE the top-level elements). Mirrors the Java/Python runners.
      // JSON fixtures ignore it.
      const outcome = extract(input, schema, schemaNode.rootless === true ? { rootless: true } : null);

      // empty flag
      expect(outcome.report.isEmpty()).toBe(expected.empty);

      // per-field states (value check)
      const actualStates = outcome.report.states();
      for (const [path, expectedState] of Object.entries(expected.states)) {
        expect(actualStates.has(path)).toBe(true);
        expect(actualStates.get(path)).toBe(expectedState as never);
      }

      // states key-set exhaustive (no extras, none missing)
      expect([...actualStates.keys()].sort()).toEqual(Object.keys(expected.states).sort());

      // Data is compared as a flat DOTTED-LEAF map (mirroring states): nested objects and
      // arrays are flattened to leaf paths (meta.score, items[0].label, tags[0], …) and every
      // leaf VALUE is asserted — including scalar-array elements and nested-object leaves.
      const actualLeaves: Record<string, unknown> = {};
      for (const [key, val] of Object.entries(outcome.data)) flattenLeaves(key, val, actualLeaves);

      for (const [path, expectedValue] of Object.entries(expected.data)) {
        expect(Object.prototype.hasOwnProperty.call(actualLeaves, path)).toBe(true);
        assertCanonical(caseName, path, expectedValue, actualLeaves[path]);
      }

      // data leaf-set exhaustive (no extras, none missing)
      expect(Object.keys(actualLeaves).sort()).toEqual(Object.keys(expected.data).sort());
    });
  }
});

File: server/java/render/src/test/java/com/metaobjects/render/extract/ExtractConformanceTest.java
package com.metaobjects.render.extract;

import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.ObjectMapper;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.junit.runners.Parameterized;

import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Path;
import java.nio.file.Paths;
import java.util.ArrayList;
import java.util.LinkedHashSet;
import java.util.List;
import java.util.Map;
import java.util.Set;
import java.util.stream.Collectors;
import java.util.stream.Stream;

import static org.junit.Assert.*;

@RunWith(Parameterized.class)
public class ExtractConformanceTest {

    private static final ObjectMapper JSON = new ObjectMapper();
    private static final Path CORPUS;
    static {
        Path p = Paths.get(System.getProperty("user.dir")).toAbsolutePath();
        while (p != null && !Files.exists(p.resolve("fixtures/extract-conformance"))) p = p.getParent();
        CORPUS = p == null ? null : p.resolve("fixtures/extract-conformance");
    }

    private final Path dir;
    public ExtractConformanceTest(String name, Path dir) { this.dir = dir; }

    @Parameterized.Parameters(name = "{0}")
    public static List<Object[]> cases() throws IOException {
        if (CORPUS == null || !Files.isDirectory(CORPUS)) return List.of();
        try (Stream<Path> s = Files.list(CORPUS)) {
            return s.filter(Files::isDirectory).sorted()
                    .map(d -> new Object[]{ d.getFileName().toString(), d })
                    .collect(Collectors.toList());
        }
    }

    @Test
    public void classificationAndCanonicalValueMatch() throws IOException {
        JsonNode schemaNode = JSON.readTree(dir.resolve("schema.json").toFile());
        ExtractSchema schema = parseSchema(schemaNode);
        String input = Files.readString(dir.resolve("input.txt"));
        JsonNode expected = JSON.readTree(dir.resolve("expected.json").toFile());

        // Optional per-fixture parse option: "rootless": true → XML response has no wrapper root
        // (a flat top-level element sequence). Other ports without this option will fail this
        // fixture — an intentional cross-port "fix me" gate.
        ExtractOptions opts = ExtractOptions.defaults();
        if (schemaNode.has("rootless") && schemaNode.get("rootless").asBoolean()) {
            opts = opts.withRootless(true);
        }
        ExtractionOutcome out = Extract.extract(input, schema, opts);

        assertEquals(dir + " empty flag", expected.get("empty").asBoolean(), out.report().isEmpty());

        JsonNode states = expected.get("states");
        states.fieldNames().forEachRemaining(path ->
            assertEquals(dir + " state[" + path + "]",
                    states.get(path).asText(), String.valueOf(out.report().states().get(path))));

        // Data is compared as a flat DOTTED-LEAF map (mirroring states): nested objects and
        // arrays are flattened to leaf paths (meta.score, items[0].label, tags[0], …) and every
        // leaf VALUE is asserted — including scalar-array elements and nested-object leaves.
        Map<String, Object> actualLeaves = new java.util.LinkedHashMap<>();
        flattenLeaves("", out.data(), actualLeaves);

        JsonNode data = expected.get("data");
        data.fieldNames().forEachRemaining(path ->
            assertCanonical(dir + " data[" + path + "]", data.get(path), actualLeaves.get(path)));

        // Exhaustive key-set checks: no missing, no extra entries in either map.
        Set<String> expectedStateKeys = new LinkedHashSet<>();
        states.fieldNames().forEachRemaining(expectedStateKeys::add);
        assertEquals(dir + " state key set", expectedStateKeys, out.report().states().keySet());

        Set<String> expectedDataKeys = new LinkedHashSet<>();
        data.fieldNames().forEachRemaining(expectedDataKeys::add);
        assertEquals(dir + " data key set", expectedDataKeys, actualLeaves.keySet());
    }

    /**
     * Flatten an assembled-data value into dotted leaf paths: maps recurse by key
     * ({@code prefix.key}), lists recurse by index ({@code prefix[i]}), and every terminal
     * scalar is recorded. Mirrors how the engine enumerates per-field states, so the data
     * leaf-set lines up 1:1 with the states leaf-set (minus dropped/malformed leaves).
     */
    private static void flattenLeaves(String prefix, Object value, Map<String, Object> out) {
        if (value instanceof Map<?, ?> m) {
            for (Map.Entry<?, ?> e : m.entrySet()) {
                String key = prefix.isEmpty() ? String.valueOf(e.getKey()) : prefix + "." + e.getKey();
                flattenLeaves(key, e.getValue(), out);
            }
        } else if (value instanceof List<?> list) {
            for (int i = 0; i < list.size(); i++) {
                flattenLeaves(prefix + "[" + i + "]", list.get(i), out);
            }
        } else {
            out.put(prefix, value);
        }
    }

    /** Canonical leaf comparison: numbers within 1e-9 tolerance; everything else string-equal. */
    private static void assertCanonical(String msg, JsonNode expected, Object actual) {
        if (expected.isNumber() && actual instanceof Number n) {
            assertEquals(msg, expected.asDouble(), n.doubleValue(), 1e-9);
        } else {
            assertEquals(msg, expected.asText(), String.valueOf(actual));
        }
    }

    private static ExtractSchema parseSchema(JsonNode n) {
        Format fmt = Format.valueOf(n.get("format").asText());
        String root = n.get("rootName").asText();
        List<FieldSpec> fields = new ArrayList<>();
        for (JsonNode f : n.get("fields")) fields.add(parseField(f));
        return new ExtractSchema(fmt, root, fields);
    }

    private static FieldSpec parseField(JsonNode f) {
        String name = f.get("name").asText();
        FieldKind kind = FieldKind.valueOf(f.get("kind").asText());
        boolean req = f.has("required") && f.get("required").asBoolean();
        if (kind == FieldKind.ENUM) {
            List<String> vals = new ArrayList<>();
            if (f.has("enumValues")) f.get("enumValues").forEach(v -> vals.add(v.asText()));
            Map<String, String> aliases = new java.util.LinkedHashMap<>();
            if (f.has("enumAlias"))
                f.get("enumAlias").fields().forEachRemaining(e -> aliases.put(e.getKey(), e.getValue().asText()));
            // FR-011: optional coerceDefault / normalize / default keys.
            String coerceDefault = f.has("coerceDefault") ? f.get("coerceDefault").asText() : null;
            String normalize = parseNormalize(f.has("normalize") ? f.get("normalize").asText() : null);
            String defaultValue = f.has("default") ? f.get("default").asText() : null;
            // Phase B (array-of-enum): kind:"ENUM" + array:true → List<enum>, each element
            // coerced through the enum pipeline and classified by indexed path.
            boolean array = f.has("array") && f.get("array").asBoolean();
            return array
                    ? FieldSpec.enumArray(name, req, vals, aliases, coerceDefault, normalize, defaultValue)
                    : FieldSpec.enumField(name, req, vals, aliases, coerceDefault, normalize, defaultValue);
        }
        if (kind == FieldKind.OBJECT) {
            boolean array = f.has("array") && f.get("array").asBoolean();
            ExtractSchema nested = null;
            if (f.has("fields")) {
                List<FieldSpec> childSpecs = new ArrayList<>();
                for (JsonNode nf : f.get("fields")) childSpecs.add(parseField(nf));
                nested = new ExtractSchema(Format.JSON, name, childSpecs);
            }
            return FieldSpec.object(name, req, array, nested);
        }
        if (f.has("min") || f.has("max")) {
            Double min = f.has("min") ? f.get("min").asDouble() : null;
            Double max = f.has("max") ? f.get("max").asDouble() : null;
            return FieldSpec.range(name, kind, req, min, max);
        }
        // @xmlText: a scalar field that receives its element's text content (the #text sentinel).
        if (f.has("textContent") && f.get("textContent").asBoolean()) {
            return FieldSpec.textContentField(name, kind, req);
        }
        // Phase B: a scalar field may carry a generalized @default absent-fill string.
        String defaultValue = f.has("default") ? f.get("default").asText() : null;
        return FieldSpec.scalar(name, kind, req, defaultValue);
    }

    /** FR-011: parse the {@code @normalize} mode string; absent → the global default "strip". */
    private static String parseNormalize(String s) {
        if (s == null) return Normalize.DEFAULT;
        if (Normalize.NONE.equals(s) || Normalize.COLLAPSE.equals(s) || Normalize.STRIP.equals(s)) return s;
        throw new IllegalArgumentException("Unknown normalize mode: " + s);
    }
}
